api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java

```diff
@@ -0,0 +1,949 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.ndarray;
+
+import ai.djl.Device;
+import ai.djl.ndarray.internal.NDArrayEx;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import ai.djl.ndarray.types.SparseFormat;
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+
+/**
+ * A base implementation of the {@link NDArray} that does nothing. This can be used for overriding
+ * the NDArray with only part of the interface implemented.
+ */
+public interface NDArrayAdapter extends NDArray {
```

Comment:
In java convention, Adapter should be a class with empty implementation.
This should not be an interface.

Reply:
I originally had it as a class, but the PpNDArray already extends NativeResource

Reply:
Should we implement NativeResource functionally in MxNDArray and PpNDArray instead?

Reply:
I am not sure, especially as the NativeResource includes data as well. It might be more difficult that way

Reply:
We can refactor other NDArray classes except for MxNDArray, PpNDArray and make it abstract class

Reply:
PpNDArray is one of the ones this should help. Another option is we make NativeResource an interface, add a BaseNativeResource class, and keep a `NativeResource<T> handle = new BaseNativeResource(...)` as instance variables inside the MxNDArray and PpNDArray and implement NativeResource instead of extending BaseNativeResource. Then, the Adapter can be the class

Reply:
For all different cases, the NDArray either holds a Java native object (TF Tensor, Onnx Tensor, DLR Tensor) or JNI/JNA handle (MXNet, PpNDArray, PtNDArray). Maybe we can make NativeResource applies to all of them, just holds different objects...

Reply:
In a call, we looked into this a bit. In DJL, BaseXXX classes are used to provide a base implementation of an interface while XXXAdapter is used only to provide empty implementations for methods in an interface. In NDArrayAdapter, it should be used only for NDArrays that don't plan to implement much of the NDArray interface. Given this limited scope for the adapter, we think it should be fine as an interface as it is